vscode-lean4/src/utils/projectInfo.ts

```diff
@@ -0,0 +1,146 @@
+    if (path.scheme === 'file') {
+        // search parent folders for a leanpkg.toml file, or a Lake lean-toolchain file.
+        while (true) {
+            // give preference to 'lean-toolchain' files if any.
+            const leanToolchain = Uri.joinPath(path, toolchainFileName);
+            if (fs.existsSync(leanToolchain.fsPath)) {
+                return [wsFolder, path, leanToolchain];
+            }
+            else {
+                const leanPkg = Uri.joinPath(path, tomlFileName);
+                if (fs.existsSync(leanPkg.fsPath)) {
+                    return [wsFolder, path, leanPkg];
+                }
+                else if (searchUpwards) {
+                    const parent = Uri.joinPath(path, '..');
+                    if (parent === path) {
+                        // no project file found.
+                        break;
+                    }
+                    path = parent;
+                }
+                else {
+                    // don't search above a WorkspaceFolder barrier.
+                    break;
+                }
+            }
+        }
+    } else {
+        // TODO: do we care about HTTP schemes?
+        return [null, null, null];
+    }
+}
+
+// Find the lean project root for the given document and return the
+// Uri for the project root and the "version" information contained
+// in any 'lean-toolchain' or 'leanpkg.toml' file found there.
+export async function findLeanPackageVersionInfo(uri: Uri) : Promise<[Uri,string]> {
+
+    const [_, packageUri, packageFileUri] =findLeanPackageRoot(uri);
+    if (!packageUri || packageUri.scheme === 'untitled') return null;
+
+    let version = null;
+    if (packageFileUri) {
+        try {
+            version = await readLeanVersionFile(packageFileUri);
+        } catch (err) {
+            console.log(err);
+        }
+    }
+
+    return [packageUri, version];
+}
+
+// Find the 'leanpkg.toml' or 'lean-toolchain' in the given package root and
+// extract the Lean version info from it.
+export async function readLeanVersion(packageUri: Uri){
+    const toolchainFileName = 'lean-toolchain';
+    const tomlFileName = 'leanpkg.toml';
+    const leanToolchain = Uri.joinPath(packageUri, toolchainFileName);
+    if (fs.existsSync(new URL(leanToolchain.toString()))) {
+        return await readLeanVersionFile(leanToolchain);
+    }
+    else {
+        const leanPkg = Uri.joinPath(packageUri, tomlFileName);
+        if (fs.existsSync(new URL(leanPkg.toString()))) {
+            return readLeanVersionFile(leanPkg);
+        }
+    }
+    return null;
+}
+
+async function readLeanVersionFile(packageFileUri : Uri) : Promise<string> {
+    const url = new URL(packageFileUri.toString());
+    const tomlFileName = 'leanpkg.toml';
+    if (packageFileUri.path.endsWith(tomlFileName))
+    {
+        return new Promise<string>((resolve, reject) => {
+            if (fs.existsSync(url)) {
+                fs.readFile(url, { encoding: 'utf-8' }, (err, data) =>{
```

Comment:
Please stick to either async or sync.  Reading a small local file is fast enough that sync would be okay, but mixing them is not really helpful (since it could still block).  The `exists` operation is also deprecated (and shouldn't be used together with readFile due to the race condition where the file is deleted between the exists and readFile calls).

So either:
```typescript
try {
  fs.readFileSync(url, {encoding: 'utf-8'}).toString()
  // ...
} catch (err) {
  // could check for err.code === 'ENOENT'
  return null;
}
```

Or async, which is just as easy:
```typescript
try {
  (await fs.promises.readFile(url, {encoding: 'utf-8'})).toString()
  // ...
} catch (err) {
  // could check for err.code === 'ENOENT'
  return null;
}
```

I would also try to drop the duplicate `existsSync` check in `readLeanVersion`.

Reply:
Generally I agree with you but the only reason I didn't use exists() here is because it is deprecated for some reason.  In general I prefer all IO to be async as hard drives and network drives can have random very long delays depending on what's going on and I hate apps that force me to wait for them to do synchronous IO... but if you feel strongly about it I can change this to readFileSync - it will have ripple effects all over removing async behavior up the chain.